### PR TITLE
Refactored code updates and optimizations

### PR DESCRIPTION
**Pull request by askmish**
Signed-off-by: askmish <ashish.k.mishra@intel.com>

### CODE REVIEW

File: src/core/src/consensus_state.rs [hunk 1]
```diff
@@ -169,9 +168,9 @@ impl ConsensusState{
     }
 
     pub fn validator_did_claim_block(&mut self, validator_info: &ValidatorInfo, wait_certificate: &WaitCertificate, poet_settings_view: &PoetSettingsView ) -> (){
-      self.aggregate_local_mean += wait_certificate.local_mean;
+      self.aggregate_local_mean += 5.5_f64; //wait_certificate.local_mean;
```
Comment (rranjan3) [inline src/core/src/consensus_state.rs]: Remove old code rather than commenting

File: src/core/src/consensus_state.rs [hunk 2]
```diff
@@ -169,9 +168,9 @@ impl ConsensusState{
...
       self.total_block_claim_count += 1;
-      self.population_samples.push_back( PopulationSample{ wait_time: wait_certificate.wait_time , local_mean: wait_certificate.local_mean}); 
+      self.population_samples.push_back( PopulationSample{ wait_time: wait_certificate.wait_time , local_mean: 5.5_f64}); //wait_certificate.local_mean});
```
Comment (rranjan3) [inline src/core/src/consensus_state.rs]: Here also

File: src/core/src/enclave_sim.rs [hunk 3]
```diff
@@ -108,79 +120,86 @@ fn truncate_biguint_to_u64(num: &BigUint) -> u64 {
...
+pub fn initialize_wait_certificate(
+    in_serialized_prev_block_wait_certificate : String,
+    in_prev_block_id : String,
+    in_poet_block_id : String,
+    in_validator_id : &Vec<u8>,
+    ) -> u64 // duration
+{
+    let mut poet_cert_map_handle = POET_CERT_MAP.lock().unwrap();
+    if !poet_cert_map_handle.poet_block_id.is_empty() {
+        //return poet_cert_map_handle.wait_certificate.duration_id
```
Comment (rranjan3) [inline src/core/src/enclave_sim.rs]: This as well

File: src/core/src/enclave_sim.rs [hunk 4]
```diff
@@ -108,79 +120,86 @@ fn truncate_biguint_to_u64(num: &BigUint) -> u64 {
...
-        let deserialized_prev_cert = serde_json::from_str(&in_serialized_prev_wait_certificate);
-        let mut prev_wait_certificate : WaitCertificate = WaitCertificate::default();
-        if deserialized_prev_cert.is_ok(){
-            prev_wait_certificate = deserialized_prev_cert.unwrap();
-            prev_block_id = prev_wait_certificate.block_id.to_owned();
-        }
-        else {
-            prev_block_id = String::from("0");
+    if !in_serialized_prev_block_wait_certificate.is_empty() {
+        let deserialized_prev_block_wait_certificate = serde_json::from_str(&in_serialized_prev_block_wait_certificate);
```
Comment (rranjan3) [inline src/core/src/enclave_sim.rs]: Wrap around

File: src/core/src/engine.rs [hunk 5]
```diff
@@ -377,10 +372,15 @@ impl Engine for Poet2Engine {
...
 }
 
-fn get_cert_from(block:&Block) -> WaitCertificate {
-    let payload = &block.payload;
-    let payload_str = String::from_utf8(payload.to_vec()).expect("Found invalid UTF-8");
-    serde_json::from_str(&payload_str).unwrap()
+fn get_cert_from(block: &Block) -> WaitCertificate {
+    let payload = &block.payload.clone();
+    let payload_str = String::from_utf8(payload.clone());
+    let delim_index = payload.iter().position(|&i| i == 0x23).unwrap(); // # char
```
Comment (rranjan3) [inline src/core/src/engine.rs]: 0x23 - put it in some constant and use here. Would be more readable like that

File: src/core/src/service.rs [hunk 6]
```diff
@@ -171,46 +169,41 @@ impl Poet2Service {
...
+				 poet2_util::blockid_to_hex_string(chain_head.previous_id),
+				 prev_wait_certificate_sig,
+                                 &validator_id);
         } else {
-            DEFAULT_WAIT_TIME
-        };
-
-        info!("Wait time: {:?}", wait_time);
+            let payload = chain_head.payload.clone();
+            let delim_index = payload.iter().position(|&i| i == 0x23).unwrap(); // # char
```
Comment (rranjan3) [inline src/core/src/service.rs]: Put this 0x23 in some constant.

File: src/core/src/service.rs [hunk 7]
```diff
@@ -231,35 +224,31 @@ impl Poet2Service {
...
-                 5.5_f64,
-             );
+    pub fn create_consensus(&mut self, summary: Vec<u8>, chain_head: Block, wait_time : u64) -> String {
+        let mut payload = chain_head.payload.clone();
+        let mut prev_wait_certificate_sig = String::new();
 
-         serial_cert.clone()
-    }
+        if chain_head.block_num != 0_u64 { // not genesis block
+            let delim_index = payload.iter().position(|&i| i == 0x23).unwrap(); // # char
```
Comment (rranjan3) [inline src/core/src/service.rs]: Use the constant here as well

File: src/core/src/engine.rs [hunk 8]
```diff
@@ -345,23 +341,22 @@ impl Engine for Poet2Engine {
 
             if !published_at_height && time::Instant::now().duration_since(start) > wait_time {
                 info!("Timer expired -- publishing block");
+                info!("wait time was : {:?} for chain head: {:?}", wait_time, chain_head.clone());
```
Comment (rranjan3) [inline src/core/src/engine.rs]: We might not want to log the complete block here. BlockId should rather suffice. Also update the log level to debug maybe.

File: src/core/src/engine.rs [hunk 9]
```diff
@@ -295,13 +288,14 @@ impl Engine for Poet2Engine {
                             // wait_time = service.calculate_wait_time(new_chain_head.clone());
                             published_at_height = false;
                             start = time::Instant::now();
-                            unsafe {
-                                CLAIM_BLOCK_DUR = service.get_next_wait_time();
-                            }
-                            service.initialize_block(Some(new_chain_head));
+                            let chain_head_block = service.get_chain_head();
+                            wait_time = time::Duration::from_secs(service.get_wait_time(chain_head_block.clone(), &validator_id));
```
Comment (rranjan3) [inline src/core/src/engine.rs]: Is the complete scope required here? time::Duration::from_secs??

File: src/core/src/consensus_state.rs (hunk 2 above)
Comment (askmish) [reply src/core/src/consensus_state.rs]: whether local_mean will be in wait_cert or not is still under discussion. Keeping the placeholders in comment to indicate  that.

File: src/core/src/service.rs (hunk 6 above)
Comment (askmish) [reply src/core/src/service.rs]: Yes, this is happening in next iteration.

File: src/core/src/engine.rs (hunk 8 above)
Comment (askmish) [reply src/core/src/engine.rs]: We need to see the whole block for now to confirm the state of the block.

File: src/core/src/engine.rs (hunk 9 above)
Comment (askmish) [reply src/core/src/engine.rs]: why not?

Comment (rranjan3) [reply src/core/src/engine.rs]: use std::time::Duration;
^ This is already there. Why have a fully qualified name then?

**Pull request merged**